Proposed split of FFTW into FFTW and FFTW_MT

bradcray opened the pull request:
Spltting the FFTW module into a single- and multi-threaded
component is nice in that it permits the multi-threaded
version to initialize itself at module initialization time.
It also permits each file to list its own "requires"
dependence if I can get that agreed upon and committed in
time to make use of it.  Committing this for Nikhil's
purusal.

bradcray commented:
@npadmana: Would you give this a look and see what you think?

npadmana commented:
@bradcray:

I like the multi-locale support.

* In testFFTW_MT, does the usethread option now get used anywhere? If not, maybe just remove it.
* I like the auto-initialization, but maybe we want a flag to turn it off if the user wants to do something funky (like not initialize it on all machines).
* Continuing on that thought, maybe make lines 59-64 a separate function, that gets called if said flag is used. Much like you did with plan_with_nthreads.
* Should the "collective" functions take an optional locale array, again if the user wants to do something different. The last one is a little bit of feature creep (and I don't have an immediate use case), but I think it's easy. And doesn't make anything harder for the baseline user.

bradcray commented:
I like these suggestions and they match things I was musing about as well.  Let me take a stab at bullets 2-3.  For bullet 4, I was thinking about this as well and whether we should make these functions methods on a locale such that one could (and in fact, would have to) say things like     

    here.cleanup_threads()
or 
    Locales.cleanup_threads() 

The main qualm I had about all this was whether it created too big of a difference between the vanilla FFTW module and this one.  The main thing that took me down this path was the fact that this module has an init routine...

If you have immediate thoughts on this, let me know.  Otherwise, I'll take a second pass on this and see where it gets us.

thomasvandoren commented:
I like the idea of making the module side effects (i.e the auto-initialization) optional. Is there a strong reason for making this a side effect of `use FFTW_MT` as opposed to having some kind of init function that a user is expected to call?

I think having the user call an init function is more deterministic and easier to debug/understand.

npadmana commented:
I think there is something nice and compact about just having the user include "FFTW_MT" and have all their FFTW calls be multithreaded with reasonable defaults.

bradcray commented:
I thought a lot about this last night, and continue to prefer the "init by default, flag to disable" approach that we're currently discussing (and I still need to code).  By analogy, the GMP module initializes itself and I think this is a nice convenience as compared to calling gmp_init() before using any GMP types or BigInts...  I also don't see any significant downside to the current approach other than (potential) resource utilization -- i.e., it's hard for me to imagine correctness issues with this approach.

I think we'll want to continue to keep an eye on this issue as we develop more modules like this.  For example, if it becomes an issue, I could imagine having all such modules support an "autoinit" config param that takes its value from a shared global.  But I'm hoping it doesn't come to that...

bradcray commented:
@npadmana and @thomasvandoren -- here's a version that addresses the feedback.  If one of you could look at it and approve committing it, I'd appreciate it.

npadmana commented:
@bradcray : Looks good. 

On the documentation for plan_with_nthreads, maybe make two points -
* That this routine can be called multiple times, overwriting previous values.
* A todo that we might want to come back and cleanup the locale story, giving the user more flexibility if needed.

bradcray commented:
@npadmana : Done.
